app/models/sources/linode.rb

```diff
@@ -0,0 +1,161 @@
+module Sources
+  class Linode < Sources::Fog
+    class << self
+      def required_fields
+        super.merge({
+          api_key: :password,
+          root_password: :password,
+          image_name: :string,
+          flavor_name: :string,
+          kernel_name: :string,
+          data_center_name: :string
+        })
+      end
+    end
+
+    ID_TYPES = ['image', 'flavor', 'kernel', 'data_center']
+
+    # Connect to Linode to make sure our config is valid
+    def validate_config!
+      valid = false
+      begin
+        names_to_ids # Force re-evaluation of ids
+        valid = (!image_id.nil? && !flavor_id.nil? && !kernel_id.nil? && !data_center_id.nil?)
+      rescue Exception => e
+        add_error(e.message)
+      end
+      valid
+    end
+
+    private
+
+    # server_is_proxy_type?(server)
+    # given a server, determine if it is running a proxy
+    def server_is_proxy_type?(server)
+      # Fog implenentation does not give access to flavor/image/etc data
+      # So doing the most basic of checks here
+      return server.name =~ /proxy/
+   end
+
+    def connection
+      @connection ||= ::Fog::Compute.new(
+        :provider => 'Linode',
+        :linode_api_key => config['api_key']
+      )
+    end
+
+    def server_by_proxy(proxy)
+      return NoServer unless validate_config!
+      server = connection.servers.select do |s|
+        s.public_ip_address == proxy.host
+      end.first
+
+      server || NoServer
+    end
+
+    # The ID_TYPES methods all are very similar, so dynamically create all
+    # of them at once
+    ID_TYPES.each do |id_type|
+      class_eval <<-RUBY
+        # image_id()
+        # flavor_id()
+        # kernel_id()
+        # data_center_id()
+        # Get the image/flavor/kernel/data center id.
+        # If the id is undefined, translate image/flavor/data center name to id by
+        # connecting to Linode.
+        # Returns cached value if already calculated
+        # Parameters:
+        #   None
+        # Returns:
+        #   - a numeric id representing the image/flavor/kernel/data center on Linode
+        #   - nil if not found
+        def #{id_type}_id
+          key = '#{id_type}_id'
+          return config[key] if config.has_key?(key)
+          names_to_ids
+          config[key]
+        end
+
+        # retrieve_image_id()
+        # retrieve_flavor_id()
+        # retrieve_kernel_id()
+        # retrieve_data_center_id()
+        # Retrieve image/flavor/kernel/data center id from Linode
+        # If the id is not found then adds to the source's persistent error log
+        # Parameters:
+        #   None
+        # Returns:
+        #   - a numeric id representing the image/flavor/kernel/data center on Linode
+        #   - nil if not found
+        def retrieve_#{id_type}_id
+          name = config['#{id_type}_name']
+
+          if '#{id_type}' == 'image'
```

Comment:
Using class_eval was a very hacky trick that I used to define methods that were nearly identical to each other.  The retrieve_image_id method is very different from retrieve_flavor_id, retrieve_kernel_id and retrieve_data_center_id.  Break it out into its own method.  It will be more efficient and more readable.